// appium-uiautomator2-driver/build/test/functional/desired.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SETTINGS_CAPS = exports.BROWSER_CAPS = exports.SCROLL_CAPS = exports.APIDEMOS_CAPS = exports.GENERIC_CAPS = exports.APIDEMOS_SCROLL_ACTIVITY = exports.APIDEMOS_MAIN_ACTIVITY = exports.APIDEMOS_PACKAGE = void 0;
exports.amendCapabilities = amendCapabilities;
const lodash_1 = __importDefault(require("lodash"));
const uiautomator2ServerLaunchTimeout = process.env.CI ? 60000 : 20000;
const uiautomator2ServerInstallTimeout = process.env.CI ? 120000 : 20000;
const ADB_EXEC_TIMEOUT = process.env.CI ? 60000 : 20000;
// ApiDemos APK URL from GitHub releases
const API_DEMOS_APK_URL = 'https://github.com/appium/android-apidemos/releases/download/v6.0.2/ApiDemos-debug.apk';
// ApiDemos package and activity constants
exports.APIDEMOS_PACKAGE = 'io.appium.android.apis';
exports.APIDEMOS_MAIN_ACTIVITY = '.ApiDemos';
exports.APIDEMOS_SCROLL_ACTIVITY = '.view.ScrollView2';
function deepFreeze(object) {
    const propNames = Object.getOwnPropertyNames(object);
    for (const name of propNames) {
        const value = object[name];
        if (value && typeof value === 'object') {
            deepFreeze(value);
        }
    }
    return Object.freeze(object);
}
function amendCapabilities(baseCaps, ...newCaps) {
    return deepFreeze({
        alwaysMatch: lodash_1.default.cloneDeep(Object.assign({}, baseCaps.alwaysMatch, ...newCaps)),
        firstMatch: [{}],
    });
}
exports.GENERIC_CAPS = deepFreeze({
    firstMatch: [{}],
    alwaysMatch: {
        'appium:deviceName': 'Android',
        platformName: 'Android',
        'appium:uiautomator2ServerLaunchTimeout': uiautomator2ServerLaunchTimeout,
        'appium:uiautomator2ServerInstallTimeout': uiautomator2ServerInstallTimeout,
        'appium:automationName': 'uiautomator2',
        'appium:adbExecTimeout': ADB_EXEC_TIMEOUT,
        'appium:ignoreHiddenApiPolicyError': true,
    },
});
exports.APIDEMOS_CAPS = amendCapabilities(exports.GENERIC_CAPS, {
    'appium:app': API_DEMOS_APK_URL,
    'appium:appPackage': exports.APIDEMOS_PACKAGE,
    'appium:appActivity': exports.APIDEMOS_MAIN_ACTIVITY,
    'appium:disableWindowAnimation': true,
});
exports.SCROLL_CAPS = amendCapabilities(exports.GENERIC_CAPS, {
    'appium:app': API_DEMOS_APK_URL,
    'appium:appPackage': exports.APIDEMOS_PACKAGE,
    'appium:appActivity': exports.APIDEMOS_SCROLL_ACTIVITY,
});
exports.BROWSER_CAPS = amendCapabilities(exports.GENERIC_CAPS, {
    browserName: 'Chrome',
});
exports.SETTINGS_CAPS = amendCapabilities(exports.GENERIC_CAPS, {
    'appium:appPackage': 'com.android.settings',
    'appium:appActivity': '.Settings',
});
