// appium-uiautomator2-driver/build/test/functional/helpers/session.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initSession = initSession;
exports.attemptToDismissAlert = attemptToDismissAlert;
exports.deleteSession = deleteSession;
const constants_1 = require("./constants");
const logger_1 = require("../../../lib/logger");
const webdriverio_1 = require("webdriverio");
const asyncbox_1 = require("asyncbox");
const INIT_RETRIES = process.env.CI ? 2 : 1;
const ALERT_CHECK_RETRIES = 5;
const ALERT_CHECK_INTERVAL = 1000;
let driver;
async function initSession(caps, remoteOpts = {}) {
    // Create the driver
    const host = constants_1.DEFAULT_HOST;
    const port = constants_1.DEFAULT_PORT;
    const opts = Object.assign({}, remoteOpts, {
        hostname: host,
        port,
        capabilities: caps,
    });
    logger_1.log.debug(`Starting session on ${host}:${port}`);
    const sessionDriver = await (0, asyncbox_1.retry)(INIT_RETRIES, async (x) => await (0, webdriverio_1.remote)(x), opts);
    if (!sessionDriver) {
        throw new Error('Failed to create session');
    }
    driver = sessionDriver;
    attemptToDismissAlert(caps);
    await sessionDriver.setTimeout({ implicit: process.env.CI ? 30000 : 5000 });
    return sessionDriver;
}
async function attemptToDismissAlert(caps) {
    // In CI environment, the alert "System UI isn't responding" may appear due to less machine resources.
    if (process.env.CI && driver) {
        for (const btnId of ['android:id/button1', 'android:id/aerr_wait']) {
            let alertFound = false;
            await (0, asyncbox_1.retryInterval)(ALERT_CHECK_RETRIES, ALERT_CHECK_INTERVAL, async function () {
                let btn;
                try {
                    btn = await driver.$(`id=${btnId}`);
                    alertFound = true;
                }
                catch {
                    // no element found, so just finish
                    return;
                }
                logger_1.log.warn('*******************************************************');
                logger_1.log.warn('*******************************************************');
                logger_1.log.warn('*******************************************************');
                logger_1.log.warn('Alert found on session startup. Trying to dismiss alert');
                logger_1.log.warn('*******************************************************');
                logger_1.log.warn('*******************************************************');
                logger_1.log.warn('*******************************************************');
                await btn.click();
                throw new Error('Alert was found, retry');
            });
            // if an alert was ever found, try to start the activity that the session should have started
            if (alertFound && driver) {
                await driver.startActivity(caps.alwaysMatch?.['appium:appPackage'], caps.alwaysMatch?.['appium:appActivity'], caps.alwaysMatch?.['appium:appWaitPackage'], caps.alwaysMatch?.['appium:appWaitActivity']);
            }
        }
    }
}
async function deleteSession() {
    try {
        if (driver) {
            await driver.deleteSession();
        }
    }
    catch {
        // ignore
    }
    finally {
        driver = undefined;
    }
}
