// appium-uiautomator2-driver/build/test/functional/commands/general/general-e2e-specs.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const desired_1 = require("../../desired");
const session_1 = require("../../helpers/session");
const asyncbox_1 = require("asyncbox");
const chai_1 = __importStar(require("chai"));
const chai_as_promised_1 = __importDefault(require("chai-as-promised"));
chai_1.default.use(chai_as_promised_1.default);
async function expectPackageAndActivity(driver, pkg, activity, timeoutMs = 5000) {
    await (0, asyncbox_1.waitForCondition)(async () => (await driver.getCurrentPackage() === pkg)
        && (await driver.getCurrentActivity() === activity), {
        waitMs: timeoutMs,
        intervalMs: 300,
    });
}
describe('general', function () {
    let driver;
    before(async function () {
        driver = await (0, session_1.initSession)(desired_1.APIDEMOS_CAPS);
    });
    after(async function () {
        await (0, session_1.deleteSession)();
    });
    describe('startActivity', function () {
        it('should launch a new package and activity', async function () {
            const appPackage = await driver.getCurrentPackage();
            const appActivity = await driver.getCurrentActivity();
            (0, chai_1.expect)(appPackage).to.equal('io.appium.android.apis');
            (0, chai_1.expect)(appActivity).to.equal('.ApiDemos');
            const startAppPackage = 'io.appium.android.apis';
            const startAppActivity = '.view.SplitTouchView';
            await driver.startActivity(startAppPackage, startAppActivity);
            await expectPackageAndActivity(driver, startAppPackage, startAppActivity);
        });
        it('should be able to launch activity with custom intent parameter category', async function () {
            const startAppPackage = 'io.appium.android.apis';
            const startAppActivity = 'io.appium.android.apis.app.HelloWorld';
            const startIntentCategory = 'appium.android.intent.category.SAMPLE_CODE';
            await driver.startActivity(startAppPackage, startAppActivity, undefined, undefined, startIntentCategory);
            await expectPackageAndActivity(driver, startAppPackage, startAppActivity);
        });
        it('should be able to launch activity with dontStopAppOnReset = true', async function () {
            const startAppPackage = 'io.appium.android.apis';
            const startAppActivity = '.os.MorseCode';
            await driver.startActivity(startAppPackage, startAppActivity);
            await expectPackageAndActivity(driver, startAppPackage, startAppActivity);
        });
        it('should be able to launch activity with dontStopAppOnReset = false', async function () {
            const startAppPackage = 'io.appium.android.apis';
            const startAppActivity = '.os.MorseCode';
            await driver.startActivity(startAppPackage, startAppActivity);
            await expectPackageAndActivity(driver, startAppPackage, startAppActivity);
        });
    });
});
