// appium-uiautomator2-driver/build/test/functional/commands/general/mobile-command-e2e-specs.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const desired_1 = require("../../desired");
const session_1 = require("../../helpers/session");
const chai_1 = __importStar(require("chai"));
const chai_as_promised_1 = __importDefault(require("chai-as-promised"));
chai_1.default.use(chai_as_promised_1.default);
describe('mobile', function () {
    let driver;
    before(async function () {
        driver = await (0, session_1.initSession)(desired_1.APIDEMOS_CAPS);
    });
    after(async function () {
        await (0, session_1.deleteSession)();
    });
    describe('mobile:shell', function () {
        it('should call execute command without proxy error, but require relaxed security flag', async function () {
            try {
                await driver.execute('mobile: shell', { command: 'echo', args: ['hello'] });
            }
            catch (e) {
                (0, chai_1.expect)(e.message).to.match(/Potentially insecure feature 'adb_shell' has not been enabled/);
            }
        });
    });
    describe('mobile:broadcast', function () {
        it('should call broadcast', async function () {
            const output = await driver.execute('mobile: broadcast', {
                action: 'io.appium.settings.sms.read',
                extras: [['s', 'max', '10']],
            });
            (0, chai_1.expect)(output).to.include('result=-1');
        });
    });
    describe('mobile:batteryInfo', function () {
        it('should get battery info', async function () {
            const { level, state } = await driver.execute('mobile: batteryInfo', {});
            (0, chai_1.expect)(level).to.be.greaterThan(0.0);
            (0, chai_1.expect)(state).to.be.greaterThan(1);
        });
    });
});
